sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp

```diff
@@ -462,30 +626,92 @@ template <class KernelType, class KernelArgType, int DIMS> class libCMBatch {
     GroupDim[1] = (uint32_t)(GlobalSize[1] / LocalSize[1]);
     GroupDim[2] = (uint32_t)(GlobalSize[2] / LocalSize[2]);
 
-    PLACEHOLDER_UNUSED(WrappedLambda_NDITEM_3DIM);
-    // TODO : Invoke invokeLambda_NDITEM_3DIM through CM's multi-threaded
-    // kernel launching with WrappedLambda_NDITEM_3DIM and dimension info
+    ESimdCPUKernel ESimdCPU((fptrVoid)InvokeLambda_NDITEM_3DIM, GroupDim,
+                            SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_NDITEM_3DIM),
+                      WrappedLambda_NDITEM_3DIM.get());
   }
 };
 
+/// Implementation for ESIMD_CPU device interface accessing ESIMD
+/// intrinsics and LibCM functionalties requred by intrinsics
 // Intrinsics
 sycl::detail::ESIMDDeviceInterface::ESIMDDeviceInterface() {
-  reserved = nullptr;
   version = ESIMDEmuPluginInterfaceVersion;
+  reserved = nullptr;
+
+  /* From 'esimd_emu_functions_v1.h' : Start */
+  cm_barrier_ptr = cm_support::barrier;
+  cm_sbarrier_ptr = cm_support::split_barrier;
+  cm_fence_ptr = cm_support::fence;
 
-  /// TODO : Fill *_ptr fields with function pointers from CM
-  /// functions prefixed with 'cm_support'
+  sycl_get_surface_base_addr_ptr = cm_support::get_surface_base_addr;
+  __cm_emu_get_slm_ptr = cm_support::get_slm_base;
+  cm_slm_init_ptr = cm_support::init_slm;
+  /* From 'esimd_emu_functions_v1.h' : End */
+}
 
-  cm_barrier_ptr = nullptr;  /* cm_support::barrier; */
-  cm_sbarrier_ptr = nullptr; /* cm_support::split_barrier; */
-  cm_fence_ptr = nullptr;    /* cm_support::fence; */
+/// Implementation for Host Kernel Launch used by
+/// piEnqueueKernelLaunch
+template <typename T> using KernelFunc = std::function<void(T)>;
 
-  sycl_get_surface_base_addr_ptr =
-      nullptr;                    /* cm_support::get_surface_base_addr; */
-  __cm_emu_get_slm_ptr = nullptr; /* cm_support::get_slm_base; */
-  cm_slm_init_ptr = nullptr;      /* cm_support::init_slm; */
+static bool isNull(int NDims, const size_t *R) {
+  return ((0 == R[0]) && (1 > NDims || 0 == R[1]) && (2 > NDims || 0 == R[2]));
 }
 
+template <int NDims, typename ArgTy> struct InvokeImpl {
+
+  static sycl::range<NDims> get_range(const size_t *a) {
+    if constexpr (NDims == 1)
+      return sycl::range<NDims>{a[0]};
+    else if constexpr (NDims == 2)
+      return sycl::range<NDims>{a[0], a[1]};
+    else if constexpr (NDims == 3)
+      return sycl::range<NDims>{a[0], a[1], a[2]};
+    throw sycl::nd_range_error();
+  }
+
+  static void invoke(void *fptr, const sycl::range<NDims> &range) {
+    auto f = reinterpret_cast<std::function<void(const ArgTy &)> *>(fptr);
+    libCMBatch<KernelFunc<const ArgTy &>, ArgTy, NDims> CmThreading(*f);
+    CmThreading.runIterationSpace(range);
+  }
+
+  static void invoke(void *fptr, const size_t *GlobalWorkSize) {
+    sycl::range<NDims> range = get_range(GlobalWorkSize);
+    invoke(fptr, range);
+  }
+
+  static void invoke(void *fptr, const size_t *GlobalWorkOffset,
```

Comment:
```suggestion
  static void invoke(void *FPtr, const size_t *GlobalWorkOffset,
```
Please, use the same style for variables naming.

Reply:
Will do.

Reply:
Done.